sdk/nodejs/provisioning/cli-downloader.ts

```diff
@@ -0,0 +1,282 @@
+import AdmZip from "adm-zip"
+import * as crypto from "crypto"
+import envPaths from "env-paths"
+import * as fs from "fs"
+import fetch from "node-fetch"
+import * as os from "os"
+import * as path from "path"
+import * as tar from "tar"
+
+import { InitEngineSessionBinaryError } from "../common/errors/index.js"
+
+const CLI_HOST = "dl.dagger.io"
+const DAGGER_CLI_BIN_PREFIX = "dagger"
+const CACHE_DIR = path.join(
+  `${process.env.XDG_CACHE_HOME?.trim() || envPaths("", { suffix: "" }).cache}`,
+  "dagger"
+)
+
+interface CliDownloaderOptions {
+  cliVersion: string
+  archive?: {
+    checksumUrl?: string
+    url?: string
+    name?(architecture: string): string
+    extract?(archivePath: string, destinationPath: string): void
+    path?(destinationFolder: string): string
+  }
+  executableFilename?(name: string): string
+}
+
+export class CliDownloader {
+  private readonly cliVersion: string
+  private readonly archive?: CliDownloaderOptions["archive"]
+  private readonly executableFilename: (name: string) => string
+
+  constructor(options: CliDownloaderOptions) {
+    if (!options.cliVersion) {
+      throw new Error("cliVersion is not set")
```

Comment:
@TomChv I'm looking through the different error class in the `error.ts`, but unsure which would be best to use.  I was thinking of this one, `InitEngineSessionBinaryError`, but maybe we need a new one. Any recommendations?

Reply:
Hey @myty ! Using `InitEngineSessionBinaryError` make total sense IMO